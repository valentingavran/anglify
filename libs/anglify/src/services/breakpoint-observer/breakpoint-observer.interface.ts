import { InjectionToken } from '@angular/core';

export type Breakpoint = 'lg' | 'md' | 'sm' | 'xl' | 'xs';

export type BreakpointSettings = { [key in Breakpoint]: number };

export const BREAKPOINT_SETTINGS = new InjectionToken<Partial<BreakpointSettings>>('Breakpoint Config');
