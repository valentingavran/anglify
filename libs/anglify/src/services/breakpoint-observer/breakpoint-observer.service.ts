import { Inject, Injectable, Optional } from '@angular/core';
import { distinctUntilChanged, fromEvent, map, shareReplay, startWith } from 'rxjs';
import { Breakpoint, BreakpointSettings, BREAKPOINT_SETTINGS } from './breakpoint-observer.interface';

@Injectable({ providedIn: 'root' })
export class BreakpointObserverService {
  private readonly breakpoints: BreakpointSettings = {
    xl: 1920,
    lg: 1280,
    md: 960,
    sm: 600,
    xs: 0,
  };

  public constructor(@Optional() @Inject(BREAKPOINT_SETTINGS) private readonly settings?: Partial<BreakpointSettings>) {
    this.breakpoints = { ...this.breakpoints, ...this.settings };
    this.checkForCorrectConfiguration();
  }

  public readonly size$ = fromEvent(window, 'resize').pipe(
    startWith(this.computeScreenSize()),
    map(() => this.computeScreenSize()),
    distinctUntilChanged(),
    shareReplay(1)
  );

  public readonly smAndDown$ = this.size$.pipe(map(size => size === 'xs' || size === 'sm'));
  public readonly sm$ = this.size$.pipe(map(size => size === 'sm'));
  public readonly smAndUp$ = this.size$.pipe(map(size => size !== 'xs'));
  public readonly mdAndDown$ = this.size$.pipe(map(size => size === 'xs' || size === 'sm' || size === 'md'));
  public readonly md$ = this.size$.pipe(map(size => size === 'md'));
  public readonly mdAndUp$ = this.size$.pipe(map(size => size !== 'xs' && size !== 'sm'));
  public readonly lgAndDown$ = this.size$.pipe(map(size => size !== 'xl'));
  public readonly lg$ = this.size$.pipe(map(size => size === 'lg'));
  public readonly lgAndUp$ = this.size$.pipe(map(size => size === 'lg' || size === 'xl'));

  private computeScreenSize(): Breakpoint {
    const matched = Object.entries(this.breakpoints).find(([, value]) => window.matchMedia(`(min-width: ${value}px)`).matches) as
      | [Breakpoint, number]
      | undefined;
    return matched ? matched[0] : 'xs';
  }

  private checkForCorrectConfiguration(): void {
    if (
      this.breakpoints.xs < this.breakpoints.sm &&
      this.breakpoints.sm < this.breakpoints.md &&
      this.breakpoints.md < this.breakpoints.lg &&
      this.breakpoints.lg < this.breakpoints.xl
    ) {
    } else {
      throw new Error(
        'Breakpoint configuration is not correct. xs must be smaller than sm, sm must be smaller than md, md must be smaller than lg, lg must be smaller than xl'
      );
    }
  }
}
