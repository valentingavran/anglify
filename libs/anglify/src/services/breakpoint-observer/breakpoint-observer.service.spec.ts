import { TestBed } from '@angular/core/testing';
import { BREAKPOINT_SETTINGS } from './breakpoint-observer.interface';
import { BreakpointObserverService } from './breakpoint-observer.service';

beforeAll(() => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
});
describe('BreakpointObserverService', () => {
  let service: BreakpointObserverService;

  beforeEach(() => {
    TestBed.configureTestingModule({});
    service = TestBed.inject(BreakpointObserverService);
  });

  it('should be created', () => {
    expect(service).toBeTruthy();
  });

  it("should have a breakpoints object with keys 'xs', 'sm', 'md', 'lg', 'xl' in descending order by default", () => {
    // @ts-expect-error
    expect(Object.keys(service.breakpoints)).toEqual(['xl', 'lg', 'md', 'sm', 'xs']);
  });

  test('that xs < sm < md < lg < xl', () => {
    // @ts-expect-error
    expect(service.breakpoints.xs < service.breakpoints.sm).toBe(true);
    // @ts-expect-error
    expect(service.breakpoints.sm < service.breakpoints.md).toBe(true);
    // @ts-expect-error
    expect(service.breakpoints.md < service.breakpoints.lg).toBe(true);
    // @ts-expect-error
    expect(service.breakpoints.lg < service.breakpoints.xl).toBe(true);
  });
});

describe('Advanced BreakpointObserverService tests', () => {
  it('should throw error if breakpoints not in descending order', () => {
    TestBed.configureTestingModule({
      providers: [
        {
          provide: BREAKPOINT_SETTINGS,
          useValue: {
            md: 500,
          },
        },
      ],
    });
    expect(() => {
      TestBed.inject(BreakpointObserverService);
    }).toThrowErrorMatchingSnapshot();
  });
});
