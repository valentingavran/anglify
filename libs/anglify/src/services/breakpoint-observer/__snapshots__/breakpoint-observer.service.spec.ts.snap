// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Advanced BreakpointObserverService tests should throw error if breakpoints not in descending order 1`] = `"Breakpoint configuration is not correct. xs must be smaller than sm, sm must be smaller than md, md must be smaller than lg, lg must be smaller than xl"`;
